flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java

```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
 		}
 	}
 
+	public RowTypeInfo(List<TypeInformation<?>> types, List<String> fieldNames) {
+		super(Row.class, types == null ? null : types.toArray(new TypeInformation[types.size()]));
+		checkNotNull(fieldNames, "FieldNames should not be null.");
+		checkArgument(
+			types.size() == fieldNames.size(),
+			"Number of field types and names is different.");
+		checkArgument(
+			types.size() == new HashSet<>(fieldNames).size(),
+			"Field names are not unique.");
+
+		this.fieldNames = new String[fieldNames.size()];
+
+		for (int i = 0; i < fieldNames.size(); i++) {
+			this.fieldNames[i] = fieldNames.get(i);
+		}
+	}
+
+	@Override
+	public void getFlatFields(String fieldExpression, int offset, List<FlatFieldDescriptor> result) {
```

Comment:
It is almost ported from `CaseClassTypeInfo.getFlatFields()` except the field index. The `CaseClassTypeInfo` is 1-based, but `RowTypeInfo` is 0-based.